dev-tools/harness.js

```diff
@@ -0,0 +1,156 @@
+/**
+ * Random Simulation harness for testing and benchmarking purposes.
+ *
+ * Pokemon Showdown - http://pokemonshowdown.com/
+ *
+ * @license MIT
+ */
+
+'use strict';
+
+const BattleStreams = require('../.sim-dist/battle-stream');
+const Dex = require('../.sim-dist/dex');
+const PRNG = require('../.sim-dist/prng').PRNG;
+const RandomPlayerAI = require('../.sim-dist/examples/random-player-ai').RandomPlayerAI;
+const Streams = require('../.lib-dist/streams');
+
+const DEFAULT_SEED = [0x09917, 0x06924, 0x0e1c8, 0x06af0];
+const AI_OPTIONS = {move: 0.7, mega: 0.6};
+
+const FORMATS = [
+	'gen7randombattle', 'gen7randomdoublesbattle', 'gen7battlefactory', 'gen6randombattle', 'gen6battlefactory',
+	'gen5randombattle', 'gen4randombattle', 'gen3randombattle', 'gen2randombattle', 'gen1randombattle',
+];
+
+class Runner {
+	constructor(options) {
+		this.prng = new PRNG(options.seed);
+		this.format = options.format;
+		this.totalGames = options.totalGames;
+		this.all = !!options.all;
+		this.logs = !!options.logs;
+		this.silent = !this.logs || !!options.silent;
+		this.async = !!options.async;
+		this.p1options = options.p1options;
+		this.p2options = options.p2options;
+
+		this.formatIndex = 0;
+		this.numGames = 0;
+	}
+
+	async run() {
+		let games = [];
+		let format, lastFormat;
+		while ((format = this.getNextFormat())) {
+			if (this.all && lastFormat && format !== lastFormat) {
+				await Promise.all(games);
+				games = [];
+			}
+
+			let seed = this.prng.seed;
+			try {
+				const game = this.runGame(format);
+				if (!this.async) await game;
+				games.push(game);
+			} catch (e) {
+				console.error(`Error running game with seed: [${seed}]\n`, e);
```

Comment:
We've got to report the format as well in order to reproduce the failure.

Reply:
> We've got to report the format as well in order to reproduce the failure.

Very true, good catch!

> Also, it could be useful if the error would output instructions to run the harness with the required parameters?

My main motivation for this harness is two-fold: 

1. primarily for selfish reasons (ie. what motivates most of my work :P), so I dont have to work as hard to maintain a separate `profiling` branch up to date with `HEAD` to profile PS
2. because I think it could be helpful to others interested in performance. I want to add a `'BENCHMARKING.md'` or `'PERFORMANCE.md'` which would explain how to run the harness and profile the app using various tools, because thats something I have a lot of experience with, and IME the easier I make it the more likely other people will care about measuring the impact of their changes etc going forward.

Using this harness as a way to smoke out the simulator is really an indirect goal, and I'm not fully sure about exactly what that use case would entail which is why any 'support' for it at this point probably seems half baked (**suggestions very welcome**). I think my stretch goal for smoke tests would be to test every move + template + gen (though obviously not all possible combinations and interactions) to ensure none of them `throw` to give us some assurance the simulator is solid and is less likely to crash (maybe some kind of custom random team generator would be required + running the battles in parallel + it would still be slow and we'd want to have a separate test suite outside of unit tests which maybe only gets run on demand or by Travis or something?)

Really, I don't know. How could something like this be useful to you? This PR is very raw (I only really opened it at this early stage so I could provide context to Zarel about a bug that ended up being due to my own refactoring), and I'm very open to making it more broadly useful than my benchmarking focus (which admittedly isn't even apparent at this stage in the development because I havent even included any timing logic yet...)

Reply:
> How could something like this be useful to you?

A use case that comes to my mind is testing experimental OMs, which are likely to be frail. For example, running this would give us assurance that Digimon Showdown works properly :)

>  I want to add a 'BENCHMARKING.md' or 'PERFORMANCE.md' which would explain how to run the harness and profile the app using various tools

Awesome. That would be enough to satisfy my concern raised.

Reply:
OFC, I shouldn't minimize the benchmarking benefits -running benchmarks in PS has always been such a PITA-, but since it's still a WIP... I wait and see :)

Reply:
You could report the entire inputlog, honestly.